### PR TITLE
Change the encoding of Pagination for Docs API

### PR DESCRIPTION
**Pull request by EricBorczuk**
Because we are doing internal pagination (over cassandra) in the Docs API, but then are creating documents and making an extra layer of paging over that, this PR adds extra information into a DocsAPI-specific Paging State object. Essentially that object is just a Base64-encoded JSON blob with

```
{
 "documentId": XXX,
 "internalPageState": YYY
}
```

The documentId is the last document that was seen and returned to the user in the last request, and the internalPageState represents the _cassandra_ page that the document was on. So when you search using a Docs page state, it will start at the relevant C* page, and then fast-forward to just past the documentId that was last seen.

This should actually make searches better in two ways:
1) Pages will be definitely consistent/correct, even with mutliple nodes (which is a bit hard to test locally)
2) Search is expected to be a bit faster (there used to be a query with a particular page size that was done to determine the page state when the final document was mid-page, that no longer occurs)

### CODE REVIEW

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java
Comment (eribeiro) [reply restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:898]: No need the lines 893, 894, and 899 with the suggestion above.

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java
Comment (eribeiro) [reply restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:829]: oops, looks like `dropWhile` is not available java8. :(

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java
Comment (eribeiro) [reply restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:1031]: Couldn't we pass `currentPageState`. If it's `null` we can anticipate this possibility in `DocumentSearchPageState`

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java
Comment (eribeiro) [reply restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:829]: Other suggestion could be to create a new List and pass it:
```
boolean idFound = false;
List<Row> newRows = new ArrayList<>();
for (Row row : rows) {
     if (docId.equals(lastSeenId)) {
         found = true;
         continue;
     }
     if (found) {
         newRows.add(row);
     }
}
return newRows;
```

File: restapi/src/main/java/io/stargate/web/docsapi/dao/DocumentSearchPageState.java
Comment (eribeiro) [inline restapi/src/main/java/io/stargate/web/docsapi/dao/DocumentSearchPageState.java:21]:
```suggestion
    this.documentId = Strings.nullToEmpty(documentId);
    this.internalPageState = Strings.nullToEmpty(internalPageState);
```

File: restapi/src/main/java/io/stargate/web/docsapi/dao/DocumentSearchPageState.java
Comment (eribeiro) [inline restapi/src/main/java/io/stargate/web/docsapi/dao/DocumentSearchPageState.java:26]:
```suggestion
    this.documentId = Strings.nullToEmpty(documentId);
    if (internalPageStateBuf != null) {
    this.internalPageState = Base64.getEncoder().encodeToString(internalPageStateBuf.array());
    }
    else {
         this.internalPageState = ""
    }
```

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java
Comment (eribeiro) [inline restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:982]:
```suggestion
          currentPageState = (initialPagingState == null) ? null :  initialPagingState.getPageState();
```

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java
Comment (eribeiro) [reply restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:1027]: Change suggestion from line 1099.

File: restapi/src/main/java/io/stargate/web/docsapi/dao/DocumentSearchPageState.java
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/dao/DocumentSearchPageState.java:21]: JSON should be able to encode `null` values, so why complicate the processing logic?

File: restapi/src/main/java/io/stargate/web/docsapi/dao/DocumentSearchPageState.java
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/dao/DocumentSearchPageState.java:26]: If `internalPageStateBuf` is `null`, it's probably a processing mistake on the server side. I'd rather throw an `IllegalArgumentException` in that case. WDYT?

Comment (EricBorczuk) [reply restapi/src/main/java/io/stargate/web/docsapi/dao/DocumentSearchPageState.java:26]: I think that's probably best, yes - I'm going to be making all of these changes on the branch I'm currently working through btw, since it's based off this branch: https://github.com/stargate/stargate/pull/748

Comment (dimas-b) [inline restapi/src/main/java/io/stargate/web/docsapi/dao/DocumentSearchPageState.java:26]: Why not use the `Encoder.encode(ByteBuffer buffer)` method?

Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/dao/DocumentSearchPageState.java:26]: Although, it's a minor point.

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java
Comment (dimas-b) [inline restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:821]: Removing from iterators may not be efficient and may not even be supported if the source is not a plain list.

Would we use filtering instead?

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java
Comment (dimas-b) [inline restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:825]: I believe we also have to support the case when the reference ID is not found at all. Could this happen when data is deleted or inserted asynchronously?

**Review by dimas-b — CHANGES_REQUESTED** [review]

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:825]: We may need to use something like `DecoratedKey` and `IPartitioner` to be able to detect when we've advanced past the previous ID.

Comment (EricBorczuk) [reply restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:825]: Ah that is an interesting edge-case...If the "last seen" ID was deleted, then some weird behavior will be observed and it will actually skip over the entire first page of results...And if data is inserted such that the "last seen" ID is pushed to the following page, the same will happen - so it will show the last seen ID's document again.

We discussed a bit about how inserts/deletions mid-stream would affect the final result, but yes let's talk through these edge cases (and how those two `DecoratedKey` and `IPartitioner` might help)

**Comment by eribeiro** [comment]
@EricBorczuk @dimas-b I some ideas for a refactoring around pagination. I branched from https://github.com/stargate/stargate/pull/748 and code it. My idea is: if you think it's an idea worth trying then Eric could maybe pull those changes into his working PR. 🙂 So, no PR of mine for this one. Wdyt? What would be the best way of discussing this suggestions?

**Comment by dimas-b** [comment]
@eribeiro : we value your input for sure. Do you have a public version of your refactoring for us to have a look?

In this particular PR, I believe we have a fundamental problem when dealing with the second and following pages (as I mentioned in my [comment](https://github.com/stargate/stargate/pull/825#discussion_r608197343) above). I think we have to solve it on a conceptual level first, then we can start refactoring. I believe it would be easier to review functional change in one PR and then refactoring/clean-up in a follow-up PR. Would this work?

**Comment by eribeiro** [comment]
@dimas-b 

> Do you have a public version of your refactoring for us to have a look?

I can push a branch on my fork of SG. I am not leaning towards not opening a PR, because it will not be merged and it can pollute SG PR. But you think it's okay to open a "WIP-DON'T MERGE" PR, I am okay with that. 

>  fundamental problem when dealing with the second and following pages (as I mentioned in my comment above). I think we have to solve it on a conceptual level first

+1. Couldn't agree more.

> functional change in one PR and then refactoring/clean-up in a follow-up PR

The change I would like to propose sort falls in between a fundamental change and refactoring. There's not a clear division. 

> Would this work?

Sure. :) Gonna push the branch to my repo and let you know asap.

Let's discuss this fundamental problem here, right?

**Comment by dimas-b** [comment]
@EricBorczuk and I talked about the paging problem offline a bit. We'd like to merge this PR more or less "as is" (with review feedback) and then work on persistence-specific improvements to allow resuming paging mid-page in a separate PR. There's potential for users observing oddities with pages Docs API queries in the presence of concurrent deletes, until we resolve the deeper problem. We'll file a GH issue with details on that.

@eribeiro : FYI. I suppose we can consider refactoring current code after merging this PR, but it really depends on what exactly is changing :) hopefully it will make resolving the paging problem easier, but let's review when PR(s) are available.

**Comment by eribeiro** [comment]
> We'd like to merge this PR more or less "as is" (with review feedback) and then work on persistence-specific improvements to allow resuming paging mid-page in a separate PR.

Sounds great! Go ahead. 🙂 

> There's potential for users observing oddities with pages Docs API queries in the presence of concurrent deletes, until we resolve the deeper problem. We'll file a GH issue with details on that.

Cool.

File: restapi/src/main/java/io/stargate/web/docsapi/dao/DocumentSearchPageState.java
Comment (EricBorczuk) [reply restapi/src/main/java/io/stargate/web/docsapi/dao/DocumentSearchPageState.java:21]: I thought I had explained this in a comment somewhere on this PR< but to write it here as well:
JSON can encode null values, sure - but if Jackson is taking data such as
`{"documentId": "1", "internalPageState": null}`
And tries to turn that into a DocumentSearchPageState object, there is no way for it to know which constructor to use, as the `null` could be referring to a String or a ByteBuffer (there are two constructors in the class)

So I decided to represent the "missing" internal page state in JSON as the empty string - this technically does not have to be done for the `documentId`

Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/dao/DocumentSearchPageState.java:21]: Thanks for the clarification - it does make sense.

File: restapi/src/main/java/io/stargate/web/docsapi/dao/DocumentSearchPageState.java
Comment (EricBorczuk) [reply restapi/src/main/java/io/stargate/web/docsapi/dao/DocumentSearchPageState.java:26]: Oh actually it is possible for a valid DocumentSearchPageState to have an internalPageStateBuf that is `null`, since `null` means "the very start of the results"

Say you have 5 small docs that fit in a single C* page; if a user requests just one doc, then the paging state that is given back to them could be represented by `{"documentId": 1, "internalPageState": ""}` (essentially: I last saw document 1 on the first page of results) - and then when that's. sent back over the wire `internalPageStateBuf` is null, and that null is sent down to C*

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java
Comment (eribeiro) [reply restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:829]: @EricBorczuk Instead of skipping by changing the list in place wouldn't be simpler to create a copy? Just in case you didn't see this snippet above. :)

**Pull request closed without merging**
